File: video_send_recv/web-app/src/video-communication/ExampleUsage.ts
/**
 * Example usage of the Video Communication API.
 * 
 * This module demonstrates how other modules can use the video communication system.
 */

import { 
  initVideoSystem, 
  createConnection, 
  connectVideo, 
  disconnectVideo,
  setEventHandlers,
  getConnectionInfo,
  getAllConnections,
  cleanupVideoSystem,
  getDefaultConfig
} from './VideoAPI';
import { VideoConfig } from './VideoCommunicationManager';
import { IntensityAnalyzer, IntensityAlert } from './IntensityAnalyzer';
// import { IntensityMetrics } from '../types';

/**
 * Example class that demonstrates advanced video communication usage.
 */
export class VideoApplication {
  private intensityAnalyzer: IntensityAnalyzer;
  private connections: Map<string, string> = new Map(); // connectionId -> userId

  constructor() {
    this.intensityAnalyzer = new IntensityAnalyzer({
      lowThreshold: 15,
      highThreshold: 235,
      changeThreshold: 40,
      stabilityWindow: 10,
      maxHistory: 200
    });
  }

  /**
   * Initialize the video application.
   */
  async initialize(): Promise<void> {
    console.log('Initializing video application...');
    
    // Initialize video system with custom config
    const config: VideoConfig = {
      ...getDefaultConfig(),
      serverUrl: 'http://localhost:8080',
      videoConstraints: {
        width: { ideal: 640, max: 640 },
        height: { ideal: 360, max: 360 },
        frameRate: { ideal: 30, max: 30 },
        facingMode: 'user'
      },
      dataChannelConfig: {
        ordered: false,
        maxRetransmits: 0,
        protocol: 'intensity-v1'
      },
      reconnectInterval: 5000,
      maxReconnectAttempts: 3
    };
    
    initVideoSystem(config);
    
    // Set up event handlers
    this.setupEventHandlers();
    
    console.log('Video application initialized');
  }

  /**
   * Add a new user with video connection.
   */
  async addUser(userId: string): Promise<string> {
    console.log(`Adding user: ${userId}`);
    
    const connectionId = `user_${userId}_${Date.now()}`;
    
    try {
      // Create connection
      await createConnection(connectionId);
      
      // Start camera
      const stream = await navigator.mediaDevices.getUserMedia({
        video: {
          width: { ideal: 640, max: 640 },
          height: { ideal: 360, max: 360 },
          frameRate: { ideal: 30, max: 30 },
          facingMode: 'user'
        },
        audio: false
      });
      console.log(`Camera started for ${userId}`);
      
      // Connect to server
      await connectVideo(connectionId, stream);
      console.log(`Connected to server for ${userId}`);
      
      // Store mapping
      this.connections.set(connectionId, userId);
      
      return connectionId;
    } catch (error) {
      console.error(`Failed to add user ${userId}:`, error);
      throw error;
    }
  }

  /**
   * Remove a user and disconnect their video.
   */
  async removeUser(userId: string): Promise<void> {
    console.log(`Removing user: ${userId}`);
    
    // Find connection for user
    const connectionId = this.findConnectionForUser(userId);
    if (!connectionId) {
      console.warn(`No connection found for user: ${userId}`);
      return;
    }
    
    try {
      // Disconnect video
      await disconnectVideo(connectionId);
      
      // Remove from mapping
      this.connections.delete(connectionId);
      
      console.log(`User ${userId} removed successfully`);
    } catch (error) {
      console.error(`Failed to remove user ${userId}:`, error);
    }
  }

  /**
   * Get intensity analysis for a user.
   */
  getIntensityAnalysis(userId: string): any {
    const connectionId = this.findConnectionForUser(userId);
    if (!connectionId) {
      return null;
    }
    
    return this.intensityAnalyzer.getIntensityStats(connectionId);
  }

  /**
   * Get all alerts.
   */
  getAllAlerts(): IntensityAlert[] {
    return this.intensityAnalyzer.getAllAlerts();
  }

  /**
   * Get alerts for a specific user.
   */
  getUserAlerts(userId: string): IntensityAlert[] {
    const connectionId = this.findConnectionForUser(userId);
    if (!connectionId) {
      return [];
    }
    
    return this.intensityAnalyzer.getAlertsForConnection(connectionId);
  }

  /**
   * Get application statistics.
   */
  getApplicationStats(): {
    totalUsers: number;
    activeConnections: number;
    totalAlerts: number;
    connections: any[];
  } {
    const allConnections = getAllConnections();
    
    return {
      totalUsers: this.connections.size,
      activeConnections: allConnections.filter(c => c.isConnected).length,
      totalAlerts: this.intensityAnalyzer.getAllAlerts().length,
      connections: allConnections.map(conn => ({
        id: conn.id,
        isConnected: conn.isConnected,
        isStreaming: conn.isStreaming,
        lastIntensity: conn.lastIntensity?.avg_intensity || 0
      }))
    };
  }

  /**
   * Shutdown the application.
   */
  async shutdown(): Promise<void> {
    console.log('Shutting down video application...');
    
    // Disconnect all users
    for (const [, userId] of this.connections) {
      await this.removeUser(userId);
    }
    
    // Cleanup video system
    await cleanupVideoSystem();
    
    console.log('Video application shutdown complete');
  }

  private setupEventHandlers(): void {
    setEventHandlers({
      onConnectionStateChange: (connectionId, state) => {
        const userId = this.connections.get(connectionId);
        console.log(`Connection state changed for ${userId || connectionId}: ${state}`);
      },
      
      onIntensityUpdate: (connectionId, metrics) => {
        // Process intensity data
        const stats = this.intensityAnalyzer.processIntensity(connectionId, metrics);
        
        // Log significant changes
        if (stats.stability < 0.5) {
          console.warn(`Low stability detected for ${connectionId}: ${(stats.stability * 100).toFixed(1)}%`);
        }
        
        // Update alerts (stored in analyzer)
      },
      
      onStatsUpdate: (connectionId, stats) => {
        const userId = this.connections.get(connectionId);
        if (userId) {
          console.log(`Stats for ${userId}: FPS=${stats.outboundFps.toFixed(1)}, Bitrate=${(stats.outboundBitrate/1000).toFixed(1)}kbps`);
        }
      },
      
      onError: (connectionId, error) => {
        const userId = this.connections.get(connectionId);
        console.error(`Error for ${userId || connectionId}:`, error);
      },
      
      onStreamReady: (connectionId, stream) => {
        const userId = this.connections.get(connectionId);
        console.log(`Stream ready for ${userId || connectionId}:`, stream);
      }
    });
  }

  private findConnectionForUser(userId: string): string | null {
    for (const [connectionId, uid] of this.connections) {
      if (uid === userId) {
        return connectionId;
      }
    }
    return null;
  }
}

/**
 * Example usage functions.
 */
export async function exampleBasicUsage(): Promise<void> {
  console.log('=== Basic Video API Usage ===');
  
  // Initialize
  const config = getDefaultConfig();
  initVideoSystem(config);
  
  // Create connection
  const connectionId = 'example_conn';
  await createConnection(connectionId);
  
  // Set up simple event handlers
  setEventHandlers({
    onIntensityUpdate: (id, metrics) => {
      console.log(`Intensity for ${id}: ${metrics.avg_intensity.toFixed(1)}`);
    },
    onConnectionStateChange: (id, state) => {
      console.log(`Connection ${id} state: ${state}`);
    }
  });
  
  // Start camera and connect
  const stream = await navigator.mediaDevices.getUserMedia({
    video: {
      width: { ideal: 640, max: 640 },
      height: { ideal: 360, max: 360 },
      frameRate: { ideal: 30, max: 30 },
      facingMode: 'user'
    },
    audio: false
  });
  await connectVideo(connectionId, stream);
  
  // Get connection info
  const info = getConnectionInfo(connectionId);
  console.log('Connection info:', info);
  
  // Cleanup
  await disconnectVideo(connectionId);
  await cleanupVideoSystem();
}

export async function exampleAdvancedUsage(): Promise<void> {
  console.log('=== Advanced Video API Usage ===');
  
  // Create application
  const app = new VideoApplication();
  await app.initialize();
  
  // Add some users
  const users = ['user1', 'user2', 'user3'];
  
  for (const userId of users) {
    try {
      const connectionId = await app.addUser(userId);
      console.log(`Added user ${userId} with connection ${connectionId}`);
    } catch (error) {
      console.error(`Failed to add user ${userId}:`, error);
    }
  }
  
  // Wait a bit for data to accumulate
  await new Promise(resolve => setTimeout(resolve, 5000));
  
  // Get statistics
  const stats = app.getApplicationStats();
  console.log('Application stats:', stats);
  
  // Get alerts
  const alerts = app.getAllAlerts();
  console.log('Total alerts:', alerts.length);
  
  // Get intensity analysis for first user
  const analysis = app.getIntensityAnalysis('user1');
  if (analysis) {
    console.log('Intensity analysis for user1:', analysis);
  }
  
  // Shutdown
  await app.shutdown();
}

/**
 * Example of integration with React components.
 */
export class VideoComponent {
  private app: VideoApplication;
  private userId: string;
  private connectionId: string | null = null;

  constructor(userId: string) {
    this.app = new VideoApplication();
    this.userId = userId;
  }

  async initialize(): Promise<void> {
    await this.app.initialize();
  }

  async startVideo(): Promise<void> {
    if (this.connectionId) {
      console.warn('Video already started');
      return;
    }

    this.connectionId = await this.app.addUser(this.userId);
  }

  async stopVideo(): Promise<void> {
    if (!this.connectionId) {
      console.warn('Video not started');
      return;
    }

    await this.app.removeUser(this.userId);
    this.connectionId = null;
  }

  getIntensityData(): any {
    return this.app.getIntensityAnalysis(this.userId);
  }

  getAlerts(): IntensityAlert[] {
    return this.app.getUserAlerts(this.userId);
  }

  async cleanup(): Promise<void> {
    if (this.connectionId) {
      await this.stopVideo();
    }
    await this.app.shutdown();
  }
}

File: video-send-recv/server/example_usage.py
"""
Example usage of the Video Communication API.

This module demonstrates how other modules can use the video communication system.
"""

import asyncio
import logging
from typing import Dict, List
from video_api import (
    init_video_system, 
    connect_video, 
    register_intensity_handler,
    unregister_intensity_handler,
    disconnect_video,
    get_connection_info,
    get_all_connections,
    shutdown_video_system
)
from video_communication import IntensityData, VideoConfig

logger = logging.getLogger(__name__)


class IntensityAnalyzer:
    """Example class that analyzes video intensity data."""
    
    def __init__(self):
        self.intensity_history: Dict[str, List[float]] = {}
        self.alerts: List[str] = []
    
    def handle_intensity_data(self, data: IntensityData):
        """Handle incoming intensity data."""
        connection_id = data.connection_id
        
        # Store intensity history
        if connection_id not in self.intensity_history:
            self.intensity_history[connection_id] = []
        
        self.intensity_history[connection_id].append(data.average_intensity)
        
        # Keep only last 100 values
        if len(self.intensity_history[connection_id]) > 100:
            self.intensity_history[connection_id].pop(0)
        
        # Analyze intensity
        self._analyze_intensity(data)
        
        # Log the data
        logger.info(f"Connection {connection_id}: "
                   f"Current={data.current_intensity:.1f}, "
                   f"Average={data.average_intensity:.1f}, "
                   f"Frames={data.frame_count}")
    
    def _analyze_intensity(self, data: IntensityData):
        """Analyze intensity data for patterns or alerts."""
        connection_id = data.connection_id
        avg_intensity = data.average_intensity
        
        # Check for very low intensity (possible camera issue)
        if avg_intensity < 10:
            alert = f"ALERT: Very low intensity ({avg_intensity:.1f}) for {connection_id}"
            self.alerts.append(alert)
            logger.warning(alert)
        
        # Check for very high intensity (possible overexposure)
        elif avg_intensity > 240:
            alert = f"ALERT: Very high intensity ({avg_intensity:.1f}) for {connection_id}"
            self.alerts.append(alert)
            logger.warning(alert)
        
        # Check for sudden changes
        if len(self.intensity_history[connection_id]) >= 10:
            recent_avg = sum(self.intensity_history[connection_id][-10:]) / 10
            older_avg = sum(self.intensity_history[connection_id][-20:-10]) / 10
            
            if abs(recent_avg - older_avg) > 50:
                alert = f"ALERT: Sudden intensity change for {connection_id}: {older_avg:.1f} -> {recent_avg:.1f}"
                self.alerts.append(alert)
                logger.warning(alert)
    
    def get_intensity_summary(self, connection_id: str) -> Dict[str, float]:
        """Get intensity summary for a connection."""
        if connection_id not in self.intensity_history:
            return {}
        
        history = self.intensity_history[connection_id]
        if not history:
            return {}
        
        return {
            "current": history[-1],
            "average": sum(history) / len(history),
            "min": min(history),
            "max": max(history),
            "samples": len(history)
        }
    
    def get_all_alerts(self) -> List[str]:
        """Get all alerts."""
        return self.alerts.copy()
    
    def clear_alerts(self):
        """Clear all alerts."""
        self.alerts.clear()


class VideoManager:
    """Example manager class that uses the video API."""
    
    def __init__(self):
        self.analyzer = IntensityAnalyzer()
        self.connections: Dict[str, str] = {}  # connection_id -> user_id mapping
    
    async def initialize(self):
        """Initialize the video system."""
        # Configure video processing
        config = VideoConfig(
            max_samples=30,
            frame_timeout=5.0,
            log_interval=1.0
        )
        
        # Initialize the video system
        init_video_system(config)
        logger.info("Video system initialized")
    
    async def handle_connection_request(self, user_id: str, sdp: str, sdp_type: str) -> Dict[str, str]:
        """Handle a new connection request."""
        connection_id = f"user_{user_id}_{int(asyncio.get_event_loop().time() * 1000)}"
        
        # Register intensity handler for this connection
        register_intensity_handler(connection_id, self.analyzer.handle_intensity_data)
        
        # Create the connection
        answer = await connect_video(connection_id, sdp, sdp_type)
        
        # Store the mapping
        self.connections[connection_id] = user_id
        
        logger.info(f"Created connection {connection_id} for user {user_id}")
        return answer
    
    async def disconnect_user(self, user_id: str):
        """Disconnect a user."""
        # Find connection for user
        connection_id = None
        for conn_id, uid in self.connections.items():
            if uid == user_id:
                connection_id = conn_id
                break
        
        if connection_id:
            # Unregister handler
            unregister_intensity_handler(connection_id)
            
            # Disconnect
            await disconnect_video(connection_id)
            
            # Remove from mapping
            del self.connections[connection_id]
            
            logger.info(f"Disconnected user {user_id} (connection {connection_id})")
    
    def get_user_intensity_summary(self, user_id: str) -> Dict[str, float]:
        """Get intensity summary for a user."""
        # Find connection for user
        connection_id = None
        for conn_id, uid in self.connections.items():
            if uid == user_id:
                connection_id = conn_id
                break
        
        if connection_id:
            return self.analyzer.get_intensity_summary(connection_id)
        return {}
    
    def get_all_alerts(self) -> List[str]:
        """Get all intensity alerts."""
        return self.analyzer.get_all_alerts()
    
    def get_connection_stats(self) -> Dict[str, Dict]:
        """Get statistics for all connections."""
        return get_all_connections()
    
    async def shutdown(self):
        """Shutdown the video system."""
        await shutdown_video_system()
        logger.info("Video system shutdown complete")


# Example usage functions
async def example_basic_usage():
    """Example of basic video API usage."""
    logger.info("=== Basic Video API Usage ===")
    
    # Initialize the system
    config = VideoConfig(max_samples=20, frame_timeout=3.0)
    init_video_system(config)
    
    # Create a connection
    connection_id = "example_conn"
    
    # Register a simple intensity handler
    def simple_handler(data: IntensityData):
        logger.info(f"Intensity: {data.average_intensity:.1f}")
    
    register_intensity_handler(connection_id, simple_handler)
    
    # Simulate connection (in real usage, this would be called by the server)
    # answer = await connect_video(connection_id, sdp, sdp_type)
    
    logger.info("Basic usage example complete")


async def example_advanced_usage():
    """Example of advanced video API usage with manager."""
    logger.info("=== Advanced Video API Usage ===")
    
    # Create manager
    manager = VideoManager()
    await manager.initialize()
    
    # Simulate handling multiple users
    users = ["user1", "user2", "user3"]
    
    for user_id in users:
        # In real usage, this would be called with actual SDP data
        logger.info(f"Would handle connection for {user_id}")
        # answer = await manager.handle_connection_request(user_id, sdp, sdp_type)
    
    # Get statistics
    stats = manager.get_connection_stats()
    logger.info(f"Connection stats: {stats}")
    
    # Get alerts
    alerts = manager.get_all_alerts()
    logger.info(f"Alerts: {alerts}")
    
    # Shutdown
    await manager.shutdown()
    logger.info("Advanced usage example complete")


if __name__ == "__main__":
    # Configure logging
    logging.basicConfig(level=logging.INFO)
    
    # Run examples
    asyncio.run(example_basic_usage())
    asyncio.run(example_advanced_usage())
